node/src/components/consensus/highway_core/state/panorama.rs

```diff
@@ -187,4 +187,29 @@ impl<C: Context> Panorama<C> {
         }
         Ok(())
     }
+
+    // Returns a bit mask that sets bit to 1 for every fault validator at equivocator's
+    // `ValidatorIndex`.
+    pub(crate) fn faults_bit_mask(&self) -> usize {
+        let mut bit_mask = 0;
+        for (idx, _) in self.enumerate().filter(|&(_, obs)| obs.is_faulty()) {
+            bit_mask |= 1 << idx.0;
+        }
+        bit_mask
+    }
+
+    /// Returns votes' hashes from this panorama that cite naively equivocator's vote.
+    pub(crate) fn need_endorsements(&self, state: &State<C>) -> Vec<&C::Hash> {
+        let local_faults = state.panorama.faults_bit_mask();
+        self.iter_correct_hashes()
```

Comment:
This only iterates over votes _directly_ cited by us.
If e.g. `A < B < C < D` and `A`'s sender is seen as faulty by `D`, then endorsements for `B` _or_ `C` are sufficient.

You already implemented that we don't necessarily endorse the directly-cited `C` at all: It might not be its sender's first vote to cite `A`.